lib/plug/static.ex

```diff
@@ -225,13 +231,18 @@ defmodule Plug.Static do
     end
   end
 
-  defp put_cache_header(conn, _, _, _) do
+  defp put_cache_header(conn, _, _, _, _, _) do
     {:stale, conn}
   end
 
-  defp etag_for_path(file_info) do
-    file_info(size: size, mtime: mtime) = file_info
-    {size, mtime} |> :erlang.phash2() |> Integer.to_string(16)
+  defp etag_for_path(file_info, et_generation, path) do
+    case et_generation do
+      {module, function} ->
+        apply(module, function, [file_info, path])
```

Comment:
We should just pass `path` here and let the function itself look up `file_info`. The `:prim_file.read_file_info/1` lookup may be unnecessary for some functions.